File: lessons/05-custom-interfaces/services/python/compute_rectangle_area_server.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from my_robot_interfaces.srv import ComputeRectangleArea


class ComputeRectangleAreaServer(Node):
    def __init__(self):
        super().__init__("compute_rectangle_area_server")
        self.server_=self.create_service(srv_type=ComputeRectangleArea,
                                     srv_name="compute_rectangle_area",
                                     callback=self.compute_rectangle_area_callback)
        self.get_logger().info("Compute Rectangle Area Server has been started.")

    def compute_rectangle_area_callback(self,request:ComputeRectangleArea.Request, response:ComputeRectangleArea.Response):
        response.area=request.length*request.width
        self.get_logger().info(f"Received request: length={request.length}, width={request.width}.")
        self.get_logger().info(f"Sending response: area={response.area}.")
        return response 

def main(args=None):
    rclpy.init(args=args)
    node=ComputeRectangleAreaServer()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__=="__main__":
    main()

File: lessons/05-custom-interfaces/services/cpp/compute_rectangle_area_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "my_robot_interfaces/srv/compute_rectangle_area.hpp"


using namespace std::placeholders; 

class ComputeRectangleArea:public rclcpp::Node{
    public:
        ComputeRectangleArea():Node("compute_rectangle_area_server") {
            server_=this->create_service<my_robot_interfaces::srv::ComputeRectangleArea>(
                "compute_rectangle_area",
                std::bind(&ComputeRectangleArea::callbackComputeArea,this,_1,_2)
            );
            RCLCPP_INFO(this->get_logger(),"Compute Rectangle Area service has been started");
        }

    private:
        rclcpp::Service<my_robot_interfaces::srv::ComputeRectangleArea>::SharedPtr server_;
        void callbackComputeArea(const my_robot_interfaces::srv::ComputeRectangleArea::Request::SharedPtr request,
                                const my_robot_interfaces::srv::ComputeRectangleArea::Response::SharedPtr response)
                                {
                                    response->area=request->length*request->width;
                                    RCLCPP_INFO(this->get_logger(),"Area of the rectangle is: %d",(int)response->area);
                                }
};

int main(int argc, char **argv) {
    rclcpp::init(argc,argv);
    auto node=std::make_shared<ComputeRectangleArea>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}